### ISSUE TITLE
ceph: dashboard set-credentials is not retried

### ISSUE THREAD

**Issue by michas2** [rook/rook#2809]
(Already discussed that issue with lbogdan in slack.)

**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Ceph dashboard does not activate.

**Expected behavior:**

Dashboard should activate, when requested in the CephCluster object.

**How to reproduce it (minimal and precise):**

I am using v0.9.2 of rook to run ceph on a bare metal typhoon cluster.
I used the [example cluster config](https://github.com/rook/rook/blob/v0.9.2/cluster/examples/kubernetes/ceph/cluster.yaml) to create a ceph cluster.

```
k -n rook-ceph get CephCluster rook-ceph -o json | jq .spec.dashboard
{
  "enabled": true,
  "ssl": false
}
```

Now the operator logs looks like this:
```
[...]
2019-03-11 14:22:22.749404 I | exec: Running command: ceph mgr module enable dashboard --force --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/449821494
I0311 14:22:27.137337       9 controller.go:1072] scheduleOperation[lock-provision-default/test[e2e4d381-4408-11e9-b311-60a44ce5fffa]]
I0311 14:22:27.145037       9 leaderelection.go:156] attempting to acquire leader lease...
I0311 14:22:27.149666       9 leaderelection.go:178] successfully acquired lease to provision for pvc default/test
I0311 14:22:27.149718       9 controller.go:1072] scheduleOperation[provision-default/test[e2e4d381-4408-11e9-b311-60a44ce5fffa]]
2019-03-11 14:22:27.153064 I | op-provisioner: creating volume with configuration {blockPool:replicapool clusterNamespace:rook-ceph fstype:xfs dataBlockPool:}
2019-03-11 14:22:27.153090 I | exec: Running command: rbd create replicapool/pvc-e2e4d381-4408-11e9-b311-60a44ce5fffa --size 3072 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring
E0311 14:22:27.232784       9 controller.go:800] Failed to provision volume for claim "default/test" with StorageClass "rook-ceph-block": Failed to create rook block image replicapool/pvc-e2e4d381-4408-11e9-b311-60a44ce5fffa: failed to create image pvc-e2e4d381-4408-11e9-b311-60a44ce5fffa in pool replicapool of size 3221225472: Failed to complete '': exit status 2. rbd: error opening pool 'replicapool': (2) No such file or directory
. output: 
E0311 14:22:27.232872       9 goroutinemap.go:150] Operation for "provision-default/test[e2e4d381-4408-11e9-b311-60a44ce5fffa]" failed. No retries permitted until 2019-03-11 14:22:59.232842275 +0000 UTC m=+242.220334657 (durationBeforeRetry 32s). Error: "Failed to create rook block image replicapool/pvc-e2e4d381-4408-11e9-b311-60a44ce5fffa: failed to create image pvc-e2e4d381-4408-11e9-b311-60a44ce5fffa in pool replicapool of size 3221225472: Failed to complete '': exit status 2. rbd: error opening pool 'replicapool': (2) No such file or directory\n. output: "
2019-03-11 14:22:28.860769 I | op-mgr: Running command: ceph dashboard set-login-credentials admin *******
I0311 14:22:29.157693       9 leaderelection.go:198] stopped trying to renew lease to provision for pvc default/test, task failed
2019-03-11 14:22:29.497422 I | exec: no valid command found; 10 closest matches:
osd lspools {<int>}
osd crush rule list
osd versions
osd map <poolname> <objectname> {<nspace>}
osd metadata {<osdname (id|osd.id)>}
osd count-metadata <property>
osd ls-tree {<int[0-]>} <name>
osd find <osdname (id|osd.id)>
osd getcrushmap {<int[0-]>}
osd getmaxosd
Error EINVAL: invalid command
2019-03-11 14:22:29.497535 E | op-mgr: failed to enable mgr dashboard. failed to initialize dashboard. failed to set login creds. failed to set login creds on mgr. exit status 22
[...]
```
however manually setting the credentials looks fine:
```
$ kubectl -n rook-ceph-system exec rook-ceph-operator-64547cdf54-fjn8w -- ceph dashboard set-login-credentials admin admin
Username and password updated
```

The dashboard port is still not running:
```
$ k -n rook-ceph exec rook-ceph-mgr-a-5c8797779d-2j7cz -- ss -tupln
Netid  State      Recv-Q Send-Q Local Address:Port               Peer Address:Port              
tcp    LISTEN     0      5      10.2.2.98:9283                  *:*                   users:(("ceph-mgr",pid=1,fd=23))
tcp    LISTEN     0      128    10.2.2.98:6800                  *:*                   users:(("ceph-mgr",pid=1,fd=29))
```

rook-ceph-mgr runs the image as specified in the config above:
```
k -n rook-ceph get po rook-ceph-mgr-a-5c8797779d-2j7cz -o json|jq .spec.containers[0].image
"ceph/ceph:v13.2.4-20190109"
```

**Environment**:

A typhoon kubernetes cluster.

* OS (e.g. from /etc/os-release):
```
$ cat /etc/os-release 
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=2023.4.0
VERSION_ID=2023.4.0
BUILD_ID=2019-02-26-0032
PRETTY_NAME="Container Linux by CoreOS 2023.4.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```
* Kernel (e.g. `uname -a`):
`Linux node1 4.19.23-coreos-r1 #1 SMP Mon Feb 25 23:40:01 -00 2019 x86_64 AMD Opteron(tm) Processor 4234 AuthenticAMD GNU/Linux`

* Cloud provider or hardware configuration:
Bare metal. 4 nodes using /dev/sd[ef] as osds.

* Rook version (use `rook version` inside of a Rook Pod):
```
$ kubectl -n rook-ceph-system exec rook-ceph-operator-64547cdf54-fjn8w -- rook version
 rook: v0.9.2
```
* Kubernetes version (use `kubectl version`):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"archive", BuildDate:"2018-12-14T20:49:34Z", GoVersion:"go1.11.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:00:57Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
Typhoon

* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)):
```
# ceph health
HEALTH_OK
```

**Comment by leseb** [comment]
@noahdesu is it related to your current PR?

**Comment by michas2** [comment]
Could it be some timing issue or race condition?
The cluster needs quite some time until it is ready to provision pvc or provide an s3 endpoint.
Maybe it tries to run some command before some service is ready?

But that would probably not explain the "invalid command"...

**Comment by noahdesu** [comment]
@michas2 could you post the log output for the `mgr` pod? the dashboard command failure does appear to be a race condition (since it works later when the command works with the toolbox), but i think it more likely to be some other issue.

@leseb this doesn't appear to be related the version of rook/ceph being reported would seem to predate dashboard stuff, so it might be a repeat of past issues.

**Comment by michas2** [comment]
@noahdesu Here is the condensed log of the pod:
```
$ k -n rook-ceph logs -p rook-ceph-mgr-a-5c8797779d-2j7cz |wc -l
47330
$ k -n rook-ceph logs -p rook-ceph-mgr-a-5c8797779d-2j7cz |grep -v Prometheus|perl -pe 's/^2019.{35}//;s/unhandled message 0x\w+/unhandled message XXX/'|sort -u

[11/Mar/2019:14:22:29] ENGINE Bus STARTED
[11/Mar/2019:14:22:29] ENGINE Bus STARTING
[11/Mar/2019:14:22:29] ENGINE Serving on 10.2.2.14:9283
[11/Mar/2019:14:22:29] ENGINE Started monitor thread '_TimeoutMonitor'.
 ceph version 13.2.4 (b10be4d44915a4d78a8e06aa31919e74927b142e) mimic (stable), process ceph-mgr, pid 1
CherryPy Checker:
 client.0 ms_handle_reset on 10.3.109.88:6790/0
 mgr finish mon failed to return metadata for osd.0: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.1: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.2: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.3: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.4: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.5: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.6: (2) No such file or directory
 mgr finish mon failed to return metadata for osd.7: (2) No such file or directory
 mgr handle_mgr_map Activating!
 mgr handle_mgr_map I am now activating
 mgr load Constructed class from module: balancer
 mgr load Constructed class from module: dashboard
 mgr load Constructed class from module: iostat
 mgr load Constructed class from module: prometheus
 mgr load Constructed class from module: restful
 mgr load Constructed class from module: status
 mgr[py] Loading python module 'balancer'
 mgr[py] Loading python module 'dashboard'
 mgr[py] Loading python module 'hello'
 mgr[py] Loading python module 'influx'
 mgr[py] Loading python module 'iostat'
 mgr[py] Loading python module 'localpool'
 mgr[py] Loading python module 'prometheus'
 mgr[py] Loading python module 'restful'
 mgr[py] Loading python module 'selftest'
 mgr[py] Loading python module 'smart'
 mgr[py] Loading python module 'status'
 mgr[py] Loading python module 'telegraf'
 mgr[py] Loading python module 'telemetry'
 mgr[py] Loading python module 'zabbix'
 mgr[restful] server not running: no certificate configured
 mgr send_beacon active
 mgr send_beacon standby
 mgr.server send_report Giving up on OSDs that haven't reported yet, sending potentially incomplete PG state to mon
 mgr.server send_report Not sending PG status to monitor yet, waiting for OSDs
 ms_deliver_dispatch: unhandled message XXX mgrreport(mon.d +53-0 packed 798 daemon_metrics=1) v6 from mon.3 10.2.1.56:0/1
 ms_deliver_dispatch: unhandled message XXX mgrreport(mon.d +53-0 packed 798) v6 from mon.3 10.2.1.56:0/1
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.0 +54-0 packed 742 daemon_metrics=2) v6 from osd.0 10.2.3.18:6800/24851
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.0 +54-0 packed 742) v6 from osd.0 10.2.3.18:6800/24851
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.1 +54-0 packed 742 daemon_metrics=2) v6 from osd.1 10.2.2.16:6800/14289
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.2 +54-0 packed 742 daemon_metrics=2) v6 from osd.2 10.2.0.16:6800/22573
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.3 +54-0 packed 742 daemon_metrics=2) v6 from osd.3 10.2.1.13:6800/12527
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.4 +54-0 packed 742 daemon_metrics=2) v6 from osd.4 10.2.0.17:6800/22660
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.5 +54-0 packed 742 daemon_metrics=2) v6 from osd.5 10.2.1.12:6800/12526
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.6 +54-0 packed 742 daemon_metrics=2) v6 from osd.6 10.2.3.17:6800/24849
 ms_deliver_dispatch: unhandled message XXX mgrreport(osd.7 +54-0 packed 742 daemon_metrics=2) v6 from osd.7 10.2.2.17:6800/14280
 pidfile_write: ignore empty --pid-file
The Application mounted at '' has an empty config.
```

**Comment by noahdesu** [comment]
thanks @michas2 could you post the whole log? you can dump it to something like https://paste.fedoraproject.org/. Also, does the full log also cover the time when you setup things and observed the dashboard configuration failure (command not found)?

**Comment by michas2** [comment]
Not sure what kind of details you are looking for, but here is the huge boring mgr log:
https://paste.fedoraproject.org/paste/7SCnXT8mK~ujZc7ScEoKsA 

Here is also the log of the operator showing the error:
https://paste.fedoraproject.org/paste/bPO1JycCWllqO6vmkQCQTw

**Comment by noahdesu** [comment]
You're right, that is a pretty boring log :) The error in the operator log definitely looks like a race condition in which there is a delay after the manager deployment starts in which the operator attempts to configure the dashboard the manager process hasn't yet fully setup/paired with the monitors. Maybe @travisn can confirm that this is possible.

Regarding the other issue you mentioned with the dashboard port. Are you able to access the dashboard after exposing the dashboard service `kubectl create -f dashboard-external.yaml` per the instructions here https://rook.io/docs/rook/v0.9/ceph-dashboard.html?

**Comment by michas2** [comment]
Not sure if it really is a race condition. The error message looks more like a typo in the command.

There actually is no "other" issue.
```
$ k -n rook-ceph get po rook-ceph-mgr-a-5c8797779d-2j7cz -o json|jq .spec.containers[0].ports
[
  {
    "containerPort": 6800,
    "name": "mgr",
    "protocol": "TCP"
  },
  {
    "containerPort": 9283,
    "name": "http-metrics",
    "protocol": "TCP"
  },
  {
    "containerPort": 8443,
    "name": "dashboard",
    "protocol": "TCP"
  }
]
```
The container declares the dashboard port, but the binary inside does not serve the dashboard. This is probably because of the `failed to enable mgr dashboard` message in the operator log.

The only question is: Why did the operator fail to enable the dashboard?

**Comment by noahdesu** [comment]
> On Mar 12, 2019, at 18:52, Michael Schnupp <notifications@github.com> wrote:
> 
> Not sure if it really is a race condition. The error message looks more like a typo in the command.
> 
It’s true a typo would produce this error, but this same error will also be produced if the manager process has not yet announced/published its commands on startup (including the dashboard commands) before the time a client attempts to configure the dashboard. The dashboard is generally working in 0.9 release, so a race condition seems like the most probable scenario.
> There actually is no "other" issue.
> 
> $ k -n rook-ceph get po rook-ceph-mgr-a-5c8797779d-2j7cz -o json|jq .spec.containers[0].ports
> [
>   {
>     "containerPort": 6800,
>     "name": "mgr",
>     "protocol": "TCP"
>   },
>   {
>     "containerPort": 9283,
>     "name": "http-metrics",
>     "protocol": "TCP"
>   },
>   {
>     "containerPort": 8443,
>     "name": "dashboard",
>     "protocol": "TCP"
>   }
> ]
> The container declares the dashboard port, but the binary inside does not serve the dashboard. This is probably because of the failed to enable mgr dashboard message in the operator log.
> 
> The only question is: Why did the operator fail to enable the dashboard?
> 
I definitely sympathize w using ss to examine the listening ports, but it would be helpful to know if the dashboard is working per the instructions / creating the dashboard service.

**Comment by michas2** [comment]
The "instructions" simply tell to set `spec.dashboard.enabled=true` to enable the dashboard inside of the container and then to "connect" it to the outside using services.
If there is no dashboard running inside the container, then the service cannot pull some dashboard out of thin air.

For debugging I already used `kubectl port-forward` to access the service and the pod directly and did not get an answer. Therefore I looked inside the container and noticed that the dashboard is not running.
According to the logs the operator tried to enable it but failed to do so. - As far as I understand I cannot do much more than setting `spec.dashboard.enabled=true` and hope the operator will bring things up.

**Comment by michas2** [comment]
I now tried to manually run the steps in the code inside the operator pod:
```
# ceph dashboard set-login-credentials admin admin
# ceph mgr module disable dashboard
# ceph mgr module enable dashboard
```
The logs (as expected) noticed this. But the dashboard still does not come up.

I noticed this line:
```
2019-03-13 08:50:53.197 7f022a4e8700  1 mgr[restful] server not running: no certificate configured
```
I deactivated ssl als already written in the report (because my ingress controller already takes care of it), maybe this is why the dashboard is unhappy?

**Comment by michas2** [comment]
Oh, seems like the error message in the operator is unrelated to the dashboard. I tried to create a pvc (default/test) at the same time before the cluster was ready. Therefore the operator probably tried to access osds which were not available, yet. - Maybe we could see that as a tiny (unrelated) bug, but it does not help with the dashboard.

**Comment by lbogdan** [comment]
Did you also disable the dashboard SSL?
```sh
ceph config set mgr mgr/dashboard/ssl false
```
(see http://docs.ceph.com/docs/mimic/mgr/dashboard/#ssl-tls-support)

**Comment by michas2** [comment]
@lbogdan Bingo. That was exactly the missing piece!
As soon as I set that option, port 8080 comes up serving the dashboard.

Looks like rook does not set this option and also does not wire the correct port:
https://github.com/rook/rook/blob/v0.9.2/pkg/operator/ceph/cluster/mgr/mgr.go#L96
(But the master version of that file seems to handle things differently.)

**Comment by michas2** [comment]
I just reinstalled my cluster and double checked the dashboard issue:
```
$ perl -wpe 's|deviceFilter:$|deviceFilter: "^sd[ef]"\n    devices: [ {name: "sde"}, {name: "sdf"} ]|;s|dashboard:$|dashboard:\n    ssl: false\n    port: 8080|' rook/cluster/examples/kubernetes/ceph/cluster.yaml | kubectl create -f -
$ kubectl -n rook-ceph-system exec rook-ceph-operator-7bb6d98ff4-29fj9 -- ceph dashboard set-login-credentials admin admin
no valid command found; 10 closest matches:
[...]
$ #wait a bit
$ kubectl -n rook-ceph-system exec "$operator" -- ceph dashboard set-login-credentials admin admin
Username and password updated
$
```
Indeed I need to wait quite a bit until `dashboard set-login-credentials` is working.

**Comment by noahdesu** [comment]
@michas2 I'm glad you have a working dashboard now! The issue you are seeing is related to changes we've made recently related to how ports/ssl settings are handled, but this seems like a new issue. Ideally you shouldn't need to issue any ceph commands directly (e.g. settings ssl to false or updating username/passwords).

**Comment by noahdesu** [comment]
@michas2 What does `ss` have to say about your ports now that it's working?

**Comment by michas2** [comment]
```
$ k -n rook-ceph exec rook-ceph-mgr-a-7bd667c7df-8vfpz -- ss -tupln
Netid  State      Recv-Q Send-Q Local Address:Port               Peer Address:Port              
tcp    LISTEN     0      5      10.2.1.18:8080                  *:*                   users:(("ceph-mgr",pid=1,fd=48))
tcp    LISTEN     0      128       *:6800                  *:*                   users:(("ceph-mgr",pid=1,fd=28))
tcp    LISTEN     0      5      10.2.1.18:9283                  *:*                   users:(("ceph-mgr",pid=1,fd=22))
```
The 8080 port comes up as soon as you do `ceph config set mgr mgr/dashboard/ssl false`.
Then you need to expose that port in your pod:
```
$ k -n rook-ceph get po rook-ceph-mgr-a-7bd667c7df-8vfpz -o json|jq .spec.containers[0].ports
[
  {
    "containerPort": 6800,
    "name": "mgr",
    "protocol": "TCP"
  },
  {
    "containerPort": 9283,
    "name": "http-metrics",
    "protocol": "TCP"
  },
  {
    "containerPort": 8080,
    "name": "dashboard",
    "protocol": "TCP"
  }
]
```
(I needed to set that port explicitly in my cluster config as the default is currently broken.)

**Comment by noahdesu** [comment]
@michas2 the picture of what's happening here is becoming much clearer here. Thanks a lot for all your feedback. Here is what is occuring:

The dashboard _is_ running but configuration doesn't fully complete after the failure you saw in the operator log with setting the credentials. Immediately the operator sets credentials the ports are configured. Since that doesn't happen the default 8080 port for non-SSL in Ceph is used. This is what was so confusing to me, since Rook doesn't ever use this port.

In any case, the issue is that most of the manager module operations are retried to avoid this specific issue, but this particular command isn't retried. The fix is to be conservative and retry this configuration. I believe there are also upstream issues in Ceph looking at this exact issue, but we can fix it on the Rook side for now.

**Comment by noahdesu** [comment]
Changing the title to accurately reflect the root cause here.

**Comment by michas2** [comment]
Not sure if that is really the root cause. - Simply retrying will probably not fix the situation of an activated dashboard without ssl.

In that case you probably should at least also:
* change the default port to 8080
* set `ceph config set mgr mgr/dashboard/ssl false`

Also is there a good reason why a retry is better than simply activating the dashboard only *after* the cluster is up?

**Comment by noahdesu** [comment]
> Not sure if that is really the root cause. - Simply retrying will probably not fix the situation of an activated dashboard without ssl.

Rook will automatically configure ssl (enable/disable) and the ports. But because of the failure observed when setting the credentials Rook did not complete its configuration. Please clarify your point if I'm missing something subtle. Port 8080 seems to be mentioned here a lot, but I think that is a distraction that has only arisen as a result of the error in the operator. Rook configures alternative non-8080 ports for the dashboard.

> Also is there a good reason why a retry is better than simply activating the dashboard only after the cluster is up?

It's not a matter of activating the dashboard before the cluster is up--the cluster monitors are up when we activate the dashboard. It's a matter of timing. Activating the dashboard does not instantaneously make the dashboard commands available, hence the race condition. Retrying solves this issue when this window of time is within a reasonable bound, and there are upstream ceph proposals for more robust solutions.

**Comment by michas2** [comment]
http://docs.ceph.com/docs/mimic/mgr/dashboard/#host-name-and-port
> By default, the ceph-mgr daemon hosting the dashboard (i.e., the currently active manager) will bind to TCP port 8443 or 8080 when SSL is disabled.

Therefore the pod should expose 8443 or 8080 depending on whether ssl is enabled or not.

**Comment by noahdesu** [comment]
> On the other hand configureDashboardModule also explicitly sets the port. Therefore simply using 8443 as a "default" port even without ssl should also work.

Rook will configure different ports for ssl/non-ssl scenarios:

https://github.com/rook/rook/blob/e4d42eece1821a39611d90a9be0ff6f4e4f0406f/pkg/operator/ceph/cluster/mgr/dashboard.go#L39

Changes to the dashboard defaults would need to be done on the Ceph side outside of Rook.

**Comment by michas2** [comment]
Actually no: https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/mgr/config.go#L48-L54

The code seems to only look at the ceph version not if ssl is requested.

**Comment by noahdesu** [comment]
I think there is some confusion (at least for me!) between what the code seems to imply and what the docs state are supported. Here are the docs:

```
Luminous: Port 7000 on http
Mimic and newer: Port 8443 on https
```

On Mimic and newer you should be able to turn off SSL and the port should stay the same (8443). That seems to match up with what the code is doing.

Whether or not ports should change when SSL is turned on/off is a valid topic, but I think the port should remain the same. Thoughts?

**Comment by michas2** [comment]
Whenever I see some traffic on port 80 I would expect to find http traffic there. (i.e. ssl turned off)
Whenever I see some traffic on port 443 I would expect to find https traffic there. (i.e. ssl turned on)
It would be to a very big surprise to me, if someone decides to break with that rule.  (Sometimes people do to circumvent simple firewall rules.)

Similarly on a high port 8080 somehow implies http and 8443 somehow implies https. (You can handle things differently, but it might confuse people.)

Because of that I would usually strongly advice not to use 8443 for plain http traffic. However in a kubernetes environment there will most certainly be a service connected to the port which can assign a more suitable port. (But it still feels like a hack.)

Most kubernetes networks I have seen (YMMV), used an Ingress controller to terminate tls traffic and used plain http inside of the cluster. - Therefore I would suggest using plain http (on port 8080) as the default, unless the user explicitly *activates* ssl in the cluster object.

**Comment by noahdesu** [comment]
Thanks for this. Super informative. The container that runs the dashboard also happens to run other web services, and could in principle run even more if new services were added. I believe there has been discussion about breaking some of these out into separate containers, but until that happens it seems like these services will have to divide up the port space anyway.
